File: web/src/__tests__/dsl/runtime/Executor.test.ts
import { describe, it, expect } from 'vitest'
import { Lexer } from '@/dsl/language/Lexer'
import { Parser } from '@/dsl/language/Parser'
import { Executor } from '@/dsl/runtime/Executor'
import { Environment } from '@/dsl/runtime/Environment'
import { BuiltinRegistry } from '@/dsl/runtime/BuiltinRegistry'
import type { DslValue, NumberVal, StringVal, DateVal, TimeVal, DateTimeVal } from '@/dsl/runtime/DslValue'
import {
  toDisplayString,
  numberVal,
  stringVal,
  dateVal,
  timeVal,
  dateTimeVal,
  serializeValue,
  deserializeValue,
} from '@/dsl/runtime/DslValue'

function execute(source: string, env?: Environment): DslValue {
  const tokens = new Lexer(source).tokenize()
  const directive = new Parser(tokens, source).parseDirective()
  return new Executor().execute(directive, env ?? Environment.create())
}

function display(source: string, env?: Environment): string {
  return toDisplayString(execute(source, env))
}

function envWithMockedTime(date: Date): Environment {
  return Environment.create().withMockedTime(date)
}

describe('Executor', () => {
  describe('number evaluation', () => {
    it('should evaluate integer', () => {
      const result = execute('[42]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(42)
    })

    it('should evaluate decimal number', () => {
      const result = execute('[3.14]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBeCloseTo(3.14)
    })

    it('should evaluate zero', () => {
      const result = execute('[0]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(0)
    })
  })

  describe('string evaluation', () => {
    it('should evaluate simple string', () => {
      const result = execute('[" hello "]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe(' hello ')
    })

    it('should evaluate empty string', () => {
      const result = execute('[""]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('')
    })
  })

  describe('display strings', () => {
    it('should display integer without decimal', () => {
      expect(display('[42]')).toBe('42')
    })

    it('should display decimal number', () => {
      expect(display('[3.14]')).toBe('3.14')
    })

    it('should display zero', () => {
      expect(display('[0]')).toBe('0')
    })

    it('should display string as-is', () => {
      expect(display('["hello"]')).toBe('hello')
    })

    it('should display undefined', () => {
      const result = execute('[first(list())]')
      expect(toDisplayString(result)).toBe('undefined')
    })
  })

  describe('serialization round-trips', () => {
    it('should round-trip number', () => {
      const val = numberVal(42)
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('NumberVal')
      expect((deserialized as NumberVal).value).toBe(42)
    })

    it('should round-trip decimal number', () => {
      const val = numberVal(3.14)
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('NumberVal')
      expect((deserialized as NumberVal).value).toBeCloseTo(3.14)
    })

    it('should round-trip string', () => {
      const val = stringVal('hello world')
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('StringVal')
      expect((deserialized as StringVal).value).toBe('hello world')
    })

    it('should round-trip date', () => {
      const val = dateVal('2024-06-15')
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('DateVal')
      expect((deserialized as DateVal).value).toBe('2024-06-15')
    })

    it('should round-trip time', () => {
      const val = timeVal('14:30:00')
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('TimeVal')
      expect((deserialized as TimeVal).value).toBe('14:30:00')
    })

    it('should round-trip datetime', () => {
      const val = dateTimeVal('2024-06-15T14:30:00')
      const deserialized = deserializeValue(serializeValue(val))
      expect(deserialized.kind).toBe('DateTimeVal')
      expect((deserialized as DateTimeVal).value).toBe('2024-06-15T14:30:00')
    })
  })

  describe('date/datetime/time functions', () => {
    it('should evaluate date function with mocked time', () => {
      const mocked = new Date(2025, 2, 15, 10, 30, 0) // March 15, 2025
      const env = envWithMockedTime(mocked)
      const result = execute('[once[date]]', env)
      expect(result.kind).toBe('DateVal')
      expect((result as DateVal).value).toBe('2025-03-15')
    })

    it('should evaluate time function with mocked time', () => {
      const mocked = new Date(2025, 0, 1, 8, 5, 30)
      const env = envWithMockedTime(mocked)
      const result = execute('[once[time]]', env)
      expect(result.kind).toBe('TimeVal')
      expect((result as TimeVal).value).toBe('08:05:30')
    })

    it('should evaluate datetime function with mocked time', () => {
      const mocked = new Date(2025, 11, 25, 23, 59, 59) // Dec 25, 2025
      const env = envWithMockedTime(mocked)
      const result = execute('[once[datetime]]', env)
      expect(result.kind).toBe('DateTimeVal')
      expect((result as DateTimeVal).value).toBe('2025-12-25T23:59:59')
    })

    it('should display datetime with comma separator', () => {
      const val = dateTimeVal('2025-01-15T14:30:00')
      expect(toDisplayString(val)).toBe('2025-01-15, 14:30:00')
    })
  })

  describe('character constants', () => {
    it('should evaluate qt (double quote)', () => {
      const result = execute('[qt]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('"')
    })

    it('should evaluate nl (newline)', () => {
      const result = execute('[nl]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('\n')
    })

    it('should evaluate tab', () => {
      const result = execute('[tab]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('\t')
    })

    it('should evaluate ret (carriage return)', () => {
      const result = execute('[ret]')
      expect(result.kind).toBe('StringVal')
      expect((result as StringVal).value).toBe('\r')
    })
  })

  describe('error handling', () => {
    it('should throw on unknown function', () => {
      expect(() => execute('[unknown_func()]')).toThrow()
    })

    it('should throw on unknown variable', () => {
      expect(() => execute('[undefined_var]')).toThrow()
    })
  })

  describe('arithmetic', () => {
    it('should add two numbers', () => {
      const result = execute('[add(3, 4)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(7)
    })

    it('should subtract two numbers', () => {
      const result = execute('[sub(10, 3)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(7)
    })

    it('should multiply two numbers', () => {
      const result = execute('[mul(3, 4)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(12)
    })

    it('should divide two numbers', () => {
      const result = execute('[div(10, 2)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(5)
    })

    it('should throw on division by zero', () => {
      expect(() => execute('[div(10, 0)]')).toThrow(/[Zz]ero/)
    })

    it('should compute modulo', () => {
      const result = execute('[mod(10, 3)]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(1)
    })

    it('should throw on modulo by zero', () => {
      expect(() => execute('[mod(10, 0)]')).toThrow(/[Zz]ero/)
    })
  })

  describe('nested calls', () => {
    it('should evaluate nested arithmetic', () => {
      const result = execute('[add(mul(2, 3), sub(10, 4))]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(12) // (2*3) + (10-4) = 6 + 6
    })

    it('should evaluate deeply nested calls', () => {
      const result = execute('[add(add(1, 2), add(3, 4))]')
      expect(result.kind).toBe('NumberVal')
      expect((result as NumberVal).value).toBe(10)
    })
  })

  describe('named arguments', () => {
    it('should handle named arguments in sort', () => {
      const result = execute('[sort(list(3, 1, 2), order: descending)]')
      expect(result.kind).toBe('ListVal')
      const items = (result as { kind: 'ListVal'; items: DslValue[] }).items
      expect(items.map(toDisplayString)).toEqual(['3', '2', '1'])
    })
  })

  describe('dynamic function classification', () => {
    it('should classify date as dynamic', () => {
      expect(BuiltinRegistry.isDynamic('date')).toBe(true)
    })

    it('should classify time as dynamic', () => {
      expect(BuiltinRegistry.isDynamic('time')).toBe(true)
    })

    it('should classify datetime as dynamic', () => {
      expect(BuiltinRegistry.isDynamic('datetime')).toBe(true)
    })

    it('should classify add as static', () => {
      expect(BuiltinRegistry.isDynamic('add')).toBe(false)
    })

    it('should classify sort as static', () => {
      expect(BuiltinRegistry.isDynamic('sort')).toBe(false)
    })

    it('should classify parse_date as static', () => {
      expect(BuiltinRegistry.isDynamic('parse_date')).toBe(false)
    })
  })

  describe('empty parentheses', () => {
    it('should allow function call with empty parens', () => {
      const result = execute('[list()]')
      expect(result.kind).toBe('ListVal')
    })

    it('should allow constant call without parens', () => {
      const result = execute('[ascending]')
      expect(result.kind).toBe('StringVal')
    })
  })
})

File: app/src/test/java/org/alkaline/taskbrain/dsl/runtime/ExecutorTest.kt
package org.alkaline.taskbrain.dsl.runtime

import org.alkaline.taskbrain.dsl.language.CallExpr
import org.alkaline.taskbrain.dsl.runtime.values.DateTimeVal
import org.alkaline.taskbrain.dsl.runtime.values.DateVal
import org.alkaline.taskbrain.dsl.runtime.values.DslValue
import org.alkaline.taskbrain.dsl.runtime.values.NumberVal
import org.alkaline.taskbrain.dsl.runtime.values.StringVal
import org.alkaline.taskbrain.dsl.runtime.values.TimeVal
import org.alkaline.taskbrain.dsl.language.Directive
import org.alkaline.taskbrain.dsl.language.DynamicCallAnalyzer
import org.alkaline.taskbrain.dsl.language.Lexer
import org.alkaline.taskbrain.dsl.language.NumberLiteral
import org.alkaline.taskbrain.dsl.language.ParseException
import org.alkaline.taskbrain.dsl.language.Parser
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

class ExecutorTest {

    private lateinit var executor: Executor

    @Before
    fun setUp() {
        executor = Executor()
    }

    private fun execute(source: String): DslValue {
        val tokens = Lexer(source).tokenize()
        val directive = Parser(tokens, source).parseDirective()
        return executor.execute(directive)
    }

    // region Number evaluation

    @Test
    fun `evaluates integer to NumberVal`() {
        val result = execute("[42]")

        assertTrue(result is NumberVal)
        assertEquals(42.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `evaluates decimal to NumberVal`() {
        val result = execute("[3.14]")

        assertTrue(result is NumberVal)
        assertEquals(3.14, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `evaluates zero`() {
        val result = execute("[0]")

        assertTrue(result is NumberVal)
        assertEquals(0.0, (result as NumberVal).value, 0.0)
    }

    // endregion

    // region String evaluation

    @Test
    fun `evaluates string to StringVal`() {
        val result = execute("[\"hello\"]")

        assertTrue(result is StringVal)
        assertEquals("hello", (result as StringVal).value)
    }

    @Test
    fun `evaluates empty string`() {
        val result = execute("[\"\"]")

        assertTrue(result is StringVal)
        assertEquals("", (result as StringVal).value)
    }

    @Test
    fun `evaluates string with special characters`() {
        val result = execute("[\"hello-world_123\"]")

        assertTrue(result is StringVal)
        assertEquals("hello-world_123", (result as StringVal).value)
    }

    // endregion

    // region Display string

    @Test
    fun `integer displays without decimal`() {
        val result = execute("[42]")
        assertEquals("42", result.toDisplayString())
    }

    @Test
    fun `decimal displays with decimal point`() {
        val result = execute("[3.14]")
        assertEquals("3.14", result.toDisplayString())
    }

    @Test
    fun `string displays its value`() {
        val result = execute("[\"hello\"]")
        assertEquals("hello", result.toDisplayString())
    }

    // endregion

    // region Serialization round-trip

    @Test
    fun `serializes and deserializes NumberVal`() {
        val original = NumberVal(42.0)
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serializes and deserializes decimal NumberVal`() {
        val original = NumberVal(3.14159)
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serializes and deserializes StringVal`() {
        val original = StringVal("hello world")
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serializes and deserializes empty StringVal`() {
        val original = StringVal("")
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serialization includes type name`() {
        val number = NumberVal(42.0)
        val string = StringVal("hello")

        assertEquals("number", number.serialize()["type"])
        assertEquals("string", string.serialize()["type"])
    }

    @Test
    fun `throws on unknown type during deserialization`() {
        val badMap = mapOf("type" to "unknown", "value" to "test")

        assertThrows(IllegalArgumentException::class.java) {
            DslValue.deserialize(badMap)
        }
    }

    @Test
    fun `throws on missing type during deserialization`() {
        val badMap = mapOf("value" to "test")

        assertThrows(IllegalArgumentException::class.java) {
            DslValue.deserialize(badMap)
        }
    }

    // endregion

    // region End-to-end

    @Test
    fun `end-to-end number parse execute serialize deserialize`() {
        // Parse and execute
        val result = execute("[42]")

        // Serialize
        val serialized = result.serialize()

        // Deserialize
        val restored = DslValue.deserialize(serialized)

        // Verify
        assertTrue(restored is NumberVal)
        assertEquals(42.0, (restored as NumberVal).value, 0.0)
        assertEquals("42", restored.toDisplayString())
    }

    @Test
    fun `end-to-end string parse execute serialize deserialize`() {
        // Parse and execute
        val result = execute("[\"hello world\"]")

        // Serialize
        val serialized = result.serialize()

        // Deserialize
        val restored = DslValue.deserialize(serialized)

        // Verify
        assertTrue(restored is StringVal)
        assertEquals("hello world", (restored as StringVal).value)
        assertEquals("hello world", restored.toDisplayString())
    }

    // endregion

    // region Date functions (Milestone 2)

    @Test
    fun `date returns today's date`() {
        // Phase 0c: bare temporal values require once[...] wrapper
        val result = execute("[once[date]]")

        assertTrue(result is DateVal)
        assertEquals(LocalDate.now(), (result as DateVal).value)
    }

    @Test
    fun `datetime returns current datetime`() {
        val before = LocalDateTime.now()
        // Phase 0c: bare temporal values require once[...] wrapper
        val result = execute("[once[datetime]]")
        val after = LocalDateTime.now()

        assertTrue(result is DateTimeVal)
        val dt = (result as DateTimeVal).value
        assertTrue(dt >= before && dt <= after)
    }

    @Test
    fun `time returns current time`() {
        val before = LocalTime.now()
        // Phase 0c: bare temporal values require once[...] wrapper
        val result = execute("[once[time]]")
        val after = LocalTime.now()

        assertTrue(result is TimeVal)
        val t = (result as TimeVal).value
        // Allow for second rollover
        assertTrue(t >= before.minusSeconds(1) && t <= after.plusSeconds(1))
    }

    // endregion

    // region Character constants (Milestone 2)

    @Test
    fun `qt returns quote character`() {
        val result = execute("[qt]")

        assertTrue(result is StringVal)
        assertEquals("\"", (result as StringVal).value)
    }

    @Test
    fun `nl returns newline character`() {
        val result = execute("[nl]")

        assertTrue(result is StringVal)
        assertEquals("\n", (result as StringVal).value)
    }

    @Test
    fun `tab returns tab character`() {
        val result = execute("[tab]")

        assertTrue(result is StringVal)
        assertEquals("\t", (result as StringVal).value)
    }

    @Test
    fun `ret returns carriage return character`() {
        val result = execute("[ret]")

        assertTrue(result is StringVal)
        assertEquals("\r", (result as StringVal).value)
    }

    // endregion

    // region Error handling (Milestone 2)

    @Test
    fun `throws on unknown function`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[unknown_func]")
        }
        assertTrue(exception.message!!.contains("Unknown function"))
        assertTrue(exception.message!!.contains("unknown_func"))
    }

    @Test
    fun `throws when date called with arguments`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[date 42]")
        }
        assertTrue(exception.message!!.contains("takes no arguments"))
    }

    // endregion

    // region Date value serialization (Milestone 2)

    @Test
    fun `serializes and deserializes DateVal`() {
        val original = DateVal(LocalDate.of(2026, 1, 25))
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serializes and deserializes TimeVal`() {
        val original = TimeVal(LocalTime.of(14, 30, 0))
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `serializes and deserializes DateTimeVal`() {
        val original = DateTimeVal(LocalDateTime.of(2026, 1, 25, 14, 30, 0))
        val serialized = original.serialize()
        val deserialized = DslValue.deserialize(serialized)

        assertEquals(original, deserialized)
    }

    @Test
    fun `DateVal displays as ISO date`() {
        val date = DateVal(LocalDate.of(2026, 1, 25))
        assertEquals("2026-01-25", date.toDisplayString())
    }

    @Test
    fun `TimeVal displays as ISO time`() {
        val time = TimeVal(LocalTime.of(14, 30, 0))
        assertEquals("14:30:00", time.toDisplayString())
    }

    @Test
    fun `DateTimeVal displays with comma separator`() {
        val datetime = DateTimeVal(LocalDateTime.of(2026, 1, 25, 14, 30, 0))
        assertEquals("2026-01-25, 14:30:00", datetime.toDisplayString())
    }

    // endregion

    // region End-to-end function calls (Milestone 2)

    @Test
    fun `end-to-end datetime`() {
        // Parse and execute
        val before = LocalDateTime.now()
        // Phase 0c: bare temporal values require once[...] wrapper
        val result = execute("[once[datetime]]")
        val after = LocalDateTime.now()

        // Serialize
        val serialized = result.serialize()

        // Deserialize
        val restored = DslValue.deserialize(serialized)

        // Verify
        assertTrue(restored is DateTimeVal)
        val dt = (restored as DateTimeVal).value
        assertTrue(dt >= before && dt <= after)
    }

    // endregion

    // region Dynamic function classification (Milestone 2)

    @Test
    fun `date is classified as dynamic`() {
        assertTrue(BuiltinRegistry.isDynamic("date"))
    }

    @Test
    fun `datetime is classified as dynamic`() {
        assertTrue(BuiltinRegistry.isDynamic("datetime"))
    }

    @Test
    fun `time is classified as dynamic`() {
        assertTrue(BuiltinRegistry.isDynamic("time"))
    }

    @Test
    fun `qt is classified as static`() {
        assertFalse(BuiltinRegistry.isDynamic("qt"))
    }

    @Test
    fun `unknown function is classified as static`() {
        assertFalse(BuiltinRegistry.isDynamic("unknown_func"))
    }

    @Test
    fun `expression with date contains dynamic calls`() {
        val directive = parse("[date]")
        assertTrue(DynamicCallAnalyzer.containsDynamicCalls(directive.expression))
    }

    @Test
    fun `expression with datetime contains dynamic calls`() {
        val directive = parse("[datetime]")
        assertTrue(DynamicCallAnalyzer.containsDynamicCalls(directive.expression))
    }

    @Test
    fun `expression with only static functions has no dynamic calls`() {
        val directive = parse("[qt]")
        assertFalse(DynamicCallAnalyzer.containsDynamicCalls(directive.expression))
    }

    @Test
    fun `literal expression has no dynamic calls`() {
        val directive = parse("[42]")
        assertFalse(DynamicCallAnalyzer.containsDynamicCalls(directive.expression))
    }

    @Test
    fun `string literal has no dynamic calls`() {
        val directive = parse("[\"hello\"]")
        assertFalse(DynamicCallAnalyzer.containsDynamicCalls(directive.expression))
    }

    private fun parse(source: String): Directive {
        val tokens = Lexer(source).tokenize()
        return Parser(tokens, source).parseDirective()
    }

    // endregion

    // region Arithmetic functions (Milestone 3)

    @Test
    fun `add returns sum of two numbers`() {
        val result = execute("[add(1, 2)]")

        assertTrue(result is NumberVal)
        assertEquals(3.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `add works with decimals`() {
        val result = execute("[add(1.5, 2.5)]")

        assertTrue(result is NumberVal)
        assertEquals(4.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `sub returns difference`() {
        val result = execute("[sub(5, 3)]")

        assertTrue(result is NumberVal)
        assertEquals(2.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `sub can produce negative result`() {
        val result = execute("[sub(3, 5)]")

        assertTrue(result is NumberVal)
        assertEquals(-2.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `mul returns product`() {
        val result = execute("[mul(3, 4)]")

        assertTrue(result is NumberVal)
        assertEquals(12.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `div returns quotient`() {
        val result = execute("[div(10, 4)]")

        assertTrue(result is NumberVal)
        assertEquals(2.5, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `div throws on division by zero`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[div(10, 0)]")
        }
        assertTrue(exception.message!!.contains("Division by zero"))
    }

    @Test
    fun `mod returns remainder`() {
        val result = execute("[mod(10, 3)]")

        assertTrue(result is NumberVal)
        assertEquals(1.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `mod throws on modulo by zero`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[mod(10, 0)]")
        }
        assertTrue(exception.message!!.contains("Modulo by zero"))
    }

    // endregion

    // region Nested calls (Milestone 3)

    @Test
    fun `nested arithmetic calls`() {
        val result = execute("[add(mul(2, 3), 4)]")

        assertTrue(result is NumberVal)
        assertEquals(10.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `deeply nested calls`() {
        val result = execute("[add(mul(2, 3), sub(10, 5))]")

        assertTrue(result is NumberVal)
        assertEquals(11.0, (result as NumberVal).value, 0.0)
    }

    @Test
    fun `triple nested calls`() {
        // (2 * 3) + (10 - (4 / 2)) = 6 + (10 - 2) = 6 + 8 = 14
        val result = execute("[add(mul(2, 3), sub(10, div(4, 2)))]")

        assertTrue(result is NumberVal)
        assertEquals(14.0, (result as NumberVal).value, 0.0)
    }

    // endregion

    // region Named arguments parsing (Milestone 3)

    @Test
    fun `parses named arguments`() {
        // Create a simple test function that accepts named arguments
        val directive = parse("[test(foo: 42)]")
        val expr = directive.expression as CallExpr

        assertEquals("test", expr.name)
        assertEquals(0, expr.args.size)
        assertEquals(1, expr.namedArgs.size)
        assertEquals("foo", expr.namedArgs[0].name)
        assertTrue(expr.namedArgs[0].value is NumberLiteral)
        assertEquals(42.0, (expr.namedArgs[0].value as NumberLiteral).value, 0.0)
    }

    @Test
    fun `parses mixed positional and named arguments`() {
        val directive = parse("[test(1, 2, foo: 3)]")
        val expr = directive.expression as CallExpr

        assertEquals("test", expr.name)
        assertEquals(2, expr.args.size)
        assertEquals(1, expr.namedArgs.size)

        assertEquals(1.0, (expr.args[0] as NumberLiteral).value, 0.0)
        assertEquals(2.0, (expr.args[1] as NumberLiteral).value, 0.0)
        assertEquals("foo", expr.namedArgs[0].name)
        assertEquals(3.0, (expr.namedArgs[0].value as NumberLiteral).value, 0.0)
    }

    @Test
    fun `parses multiple named arguments`() {
        val directive = parse("[test(a: 1, b: 2, c: 3)]")
        val expr = directive.expression as CallExpr

        assertEquals("test", expr.name)
        assertEquals(0, expr.args.size)
        assertEquals(3, expr.namedArgs.size)

        assertEquals("a", expr.namedArgs[0].name)
        assertEquals("b", expr.namedArgs[1].name)
        assertEquals("c", expr.namedArgs[2].name)
    }

    @Test
    fun `positional argument after named throws parse error`() {
        val exception = assertThrows(ParseException::class.java) {
            parse("[test(foo: 1, 2)]")
        }
        assertTrue(exception.message!!.contains("Positional argument cannot follow named argument"))
    }

    // endregion

    // region Arithmetic error handling (Milestone 3)

    @Test
    fun `add with wrong arity throws`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[add(1)]")
        }
        assertTrue(exception.message!!.contains("requires 2 arguments"))
    }

    @Test
    fun `add with non-number throws`() {
        val exception = assertThrows(ExecutionException::class.java) {
            execute("[add(1, \"hello\")]")
        }
        assertTrue(exception.message!!.contains("must be a number"))
    }

    @Test
    fun `arithmetic functions are classified as static`() {
        assertFalse(BuiltinRegistry.isDynamic("add"))
        assertFalse(BuiltinRegistry.isDynamic("sub"))
        assertFalse(BuiltinRegistry.isDynamic("mul"))
        assertFalse(BuiltinRegistry.isDynamic("div"))
        assertFalse(BuiltinRegistry.isDynamic("mod"))
    }

    // endregion

    // region Empty parentheses (Milestone 3)

    @Test
    fun `empty parentheses work for zero-arg functions`() {
        // Phase 0c: bare temporal values require once[...] wrapper
        val result = execute("[once[date()]]")

        assertTrue(result is DateVal)
        assertEquals(LocalDate.now(), (result as DateVal).value)
    }

    // endregion
}
